src/transformers/modeling_reformer.py

```diff
@@ -0,0 +1,1757 @@
+        elif len(vectors.shape) == 3:
+            return torch.reshape(vectors, split_dim_shape)
+        else:
+            raise ValueError("Input vector rank should be one of [3, 4], but is: {}".format(len(vectors.shape)))
+
+    def _merge_seq_length_dims(self, vectors, num_attn_heads):
+        """
+            merges factorized sequence length dims into one dim
+        """
+        batch_size = vectors.shape[0]
+        new_dim_shape = (batch_size, num_attn_heads, -1)
+
+        if len(vectors.shape) == 5:
+            return torch.reshape(vectors, new_dim_shape + (vectors.shape[-1],))
+        elif len(vectors.shape) == 4:
+            return torch.reshape(vectors, new_dim_shape)
+        else:
+            raise ValueError("Input vector rank should be one of [4, 5], but is: {}".format(len(vectors.shape)))
+
+
+class LSHSelfAttention(nn.Module, EfficientAttentionUtils):
+    def __init__(self, config):
+        super().__init__()
+        self.chunk_length = config.lsh_attn_chunk_length
+        self.num_hashes = config.num_hashes
+        self.num_buckets = config.num_buckets
+        self.num_chunks_before = config.lsh_num_chunks_before
+        self.num_chunks_after = config.lsh_num_chunks_after
+        self.hash_seed = config.hash_seed
+        self.is_decoder = config.is_decoder
+        self.max_position_embeddings = config.max_position_embeddings
+
+        self.dropout = config.lsh_attention_probs_dropout_prob
+
+        self.num_attention_heads = config.num_attention_heads
+        self.attention_head_size = config.attention_head_size
+        self.all_head_size = self.num_attention_heads * self.attention_head_size
+        self.hidden_size = config.hidden_size
+
+        # projection matrices
+        self.query_key = nn.Linear(self.hidden_size, self.all_head_size, bias=False)
+        self.value = nn.Linear(self.hidden_size, self.all_head_size, bias=False)
+
+        # save mask value here. Need fp32 and fp16 mask values
+        self.register_buffer("self_mask_value_float16", torch.tensor(-1e3))
+        self.register_buffer("self_mask_value_float32", torch.tensor(-1e5))
+        self.register_buffer("mask_value_float16", torch.tensor(-1e4))
+        self.register_buffer("mask_value_float32", torch.tensor(-1e9))
+
+    def forward(
+        self,
+        hidden_states,
+        attention_mask=None,
+        head_mask=None,
+        num_hashes=None,
+        do_output_attentions=False,
+        buckets=None,
+        **kwargs
+    ):
+        sequence_length = hidden_states.shape[1]
+        batch_size = hidden_states.shape[0]
+
+        # num hashes can optionally be overwritten by user
+        num_hashes = num_hashes if num_hashes is not None else self.num_hashes
+
+        # project hidden_states to query_key and value
+        query_key_vectors = self.query_key(hidden_states)
+        value_vectors = self.value(hidden_states)
+
+        # free memory
```

Comment:
does this actually help?! I have never had good luck with it.

Reply:
Yeah it does, since some of these tensors alone hold 1GB of memory. I played around with it using the `benchmarking.py` script